doc/compiler/ObjectLockword.md

```diff
@@ -0,0 +1,386 @@
+
+#### Unlocking, currently owns the lock
+The RC field is decremented by 1 to update information about lock nesting depth. If the RC field is now 0, the object is fully unlocked. One thing to note here is that the TID does not get cleared. The TID remains set to indicate which thread owns the reservation on the lock.
+
+#### Unlocking, does not own the lock
+An IllegalMonitorStateException exception is thrown since there is an attempt to unlock an object that the current thread does not own.
+
+### Inflated
+All locking and unlocking actions are handled in slow path VM code. If a lock no longer needs to be inflated, it reverts to Flat-Unlocked.
+
+# Current lockword format
+## Current Lockword Bit Pattern
+```
+64 bit lockword:
+| TID (56 bits) | RC/LC (4 bits) | Lrn | Res | FLC | Infl |
+32 bit lockword:
+| TID (24 bits) | RC/LC (4 bits) | Lrn | Res | FLC | Infl |
+
+ TID = Thread ID excluding lowest 8 bits
+  RC = Recursion Count
+  LC = Learning Count
+ Lrn = Learning Bit
+ Res = Reserved Bit
+ FLC = Flatlock Contention Bit
+Infl = Inflated Bit
+
+The 4 bit RC/LC field shares the 4 bits between the RC and LC in different ways depending on the lock state.
+
+Non-Learning state case:
+| TID | RC (4 bits) | Lrn | Res | FLC | Infl |
+
+Learning state case:
+| TID | RC (2 bits) | LC (2 bits) | Lrn | Res | FLC | Infl |
+
+The Inflated state will override the other bits so there is no RC or LC field anymore.
+```
+
+Compared to the original format the RC field is now 1 bit smaller to make room for the new Learning bit. Also there is a new LC or Learning Count field that partially shares bits with the RC.
+
+The TID field is largely handled the same way as before. It stores the TID (minus the lower 8 bits which are always 0) of the thread that currently owns the lock or the thread that reserved the lock. It can now also hold the TID of a thread that is trying to reserve the lock. This is connected to the new Learning state.
+
+The RC or Recursion Count field is primarily used to track nested locking. Reserved locks and Flat locks are treated the same way as before. Locks in the Learning state handle RC in the same was as lock in the Reserved state. If a lock has been locked exactly 2 times with no unlocking actions in between and the lock is currently in the Learning state, then the RC will hold a value of 2.
+
+The LC or Learning Count is new. This field only appears for locks in the Learning state. It gets incremented whenever a lock is locked by the same thread. After it reaches a certain threshold the lock will transition from the Learning state to the Reserved state.
```

Comment:
Behaviour on LC overflow in Learning would be helpful to clarify

Reply:
Same as other two comments, I'll add them in.